### ISSUE TITLE
F821 false-positive for SQLAlchemy mappings with not yet declared classes

### ISSUE THREAD

**Issue by aberres** [astral-sh/ruff#10451]
The described behavior happens since 0.3.3. It might be related to https://github.com/astral-sh/ruff/issues/10340

It is triggered throughout my SQLAlchemy models in cases where a model is not defined yet while used in  a type annotation

Minimal repro:

```python
from __future__ import annotations

from sqlalchemy.orm import DeclarativeBase, Mapped


class Base(DeclarativeBase):
    some_mapping: Mapped[list[Bar]] | None = None # F821 Undefined name `Bar`
    simplified: list[Bar] | None = None # F821 Undefined name `Bar`


class Bar:
    pass
```

I can only trigger the error when inheriting from `DeclarativeBase`. Without this inheritance or when inheriting from another class, things are fine.

**Comment by AlexWaygood** [comment]
Hi! I can't reproduce your error locally or in the Ruff playground: https://play.ruff.rs/1f9ab24b-394d-46c3-923a-ecf866705693

Can you paste a snippet that reproduces the error in the Ruff playground?

**Comment by aberres** [comment]
Does the playground have a real installation of `SQLAlchemy`?

I can only reproduce it locally with the actual `sqlalchemy.orm.DeclarativeBase`. Other classes as a base class are fine.

**Comment by AlexWaygood** [comment]
I don't think it should matter from Ruff's perspective whether `sqlalchemy` is installed into a Python environment or not, but let me check that :-)

**Comment by AlexWaygood** [comment]
I can't reproduce locally with SQLAlchemy installed either.

Does it still reproduce after running `ruff clean`?

**Comment by carl-baillargeon** [comment]
We are also hitting this issue in this project: https://github.com/arista-netdevops-community/anta

If you install the project locally with 0.3.3 you should see a bunch of F821 errors:

```bash
(.anta-dev) ~/git_projects/anta (main ✗) ruff check 
anta/models.py:273:27: F821 Undefined name `ResultOverwrite`
anta/models.py:274:18: F821 Undefined name `Filters`
anta/tests/bfd.py:40:25: F821 Undefined name `BFDPeer`
anta/tests/bfd.py:93:25: F821 Undefined name `BFDPeer`
anta/tests/connectivity.py:35:21: F821 Undefined name `Host`
anta/tests/connectivity.py:94:25: F821 Undefined name `Neighbor`
anta/tests/interfaces.py:176:26: F821 Undefined name `InterfaceState`
anta/tests/interfaces.py:535:26: F821 Undefined name `InterfaceDetail`
anta/tests/routing/bgp.py:175:32: F821 Undefined name `BgpAfi`
anta/tests/routing/bgp.py:285:32: F821 Undefined name `BgpAfi`
anta/tests/routing/bgp.py:397:32: F821 Undefined name `BgpAfi`
anta/tests/routing/bgp.py:509:25: F821 Undefined name `BgpNeighbor`
anta/tests/routing/bgp.py:581:25: F821 Undefined name `BgpPeer`
anta/tests/routing/bgp.py:653:25: F821 Undefined name `BgpPeer`
anta/tests/routing/bgp.py:719:25: F821 Undefined name `BgpPeer`
anta/tests/routing/bgp.py:785:25: F821 Undefined name `BgpPeer`
anta/tests/routing/bgp.py:846:31: F821 Undefined name `VxlanEndpoint`
anta/tests/routing/bgp.py:909:25: F821 Undefined name `BgpPeer`
anta/tests/routing/bgp.py:968:25: F821 Undefined name `BgpPeer`
anta/tests/security.py:303:28: F821 Undefined name `APISSLCertificate`
anta/tests/security.py:462:33: F821 Undefined name `IPv4ACL`
anta/tests/security.py:471:27: F821 Undefined name `IPv4ACLEntry`
anta/tests/services.py:106:27: F821 Undefined name `DnsServer`
anta/tests/services.py:159:23: F821 Undefined name `ErrDisableReason`
Found 24 errors.
```


Thanks

**Comment by charliermarsh** [comment]
This reproduces:

```python
from __future__ import annotations

from pydantic import BaseModel

 
class Input(BaseModel):

    result_overwrite: ResultOverwrite | None = None
    filters: Filters | None = None

    class ResultOverwrite(BaseModel):
        ...

    class Filters(BaseModel):
        ...

```

**Comment by AlexWaygood** [comment]
Okay, so `lint.flake8-type-checking.runtime-evaluated-base-classes` needs to be set to `["sqlalchemy.orm.DeclarativeBase"]` in order for it to reproduce

**Comment by carl-baillargeon** [comment]
> This reproduces:
> 
> ```python
> from __future__ import annotations
> 
> from pydantic import BaseModel
> 
>  
> class Input(BaseModel):
> 
>     result_overwrite: ResultOverwrite | None = None
>     filters: Filters | None = None
> 
>     class ResultOverwrite(BaseModel):
>         ...
> 
>     class Filters(BaseModel):
>         ...
> ```

Should I open a separate issue to track it?

**Comment by aberres** [comment]
@carl-baillargeon Should be the same issue due to `pydantic.BaseModel` being listed in `lint.flake8-type-checking.runtime-evaluated-base-classes`.

**Comment by AlexWaygood** [comment]
Confirmed that this (unsurprisingly) bisects to https://github.com/astral-sh/ruff/commit/704fefc7ab9c84c7c10b9dfaaf68a4d071ae62d3. Sorry for the regression!

**Comment by insilications** [comment]
> @carl-baillargeon Should be the same issue due to `pydantic.BaseModel` being listed in `lint.flake8-type-checking.runtime-evaluated-base-classes`.

Confirmed.

**Comment by adrianovieira** [comment]
Hi folks!

I am also getting this issue here.

```bash
F821 Undefined name `User`
  --> src/api/models/orm/address.py:18:27
   |
17 |     user_id: Mapped[int] = mapped_column(ForeignKey("user.id"))
18 |     users: Mapped[list["User"]] = relationship(back_populates="addresses")
   |                         ^^^^
```

- `address.py`:   
    ```python
    # address.py
    from sqlalchemy import ForeignKey
    from sqlalchemy.orm import DeclarativeBase, Mapped, mapped_column, relationship

    # Shoudn't import 'User'  =>  Error: ... (most likely due to a circular import)
    # from .user import User as User

    class Address(DeclarativeBase):
        __tablename__ = "address"

        user_id:  Mapped[int] = mapped_column(ForeignKey("atleta.id"))
        users:  Mapped[list["User"]] = relationship(back_populates="addresses")

        ...

    ```


<details>
<summary>User</summary>

```python
# user.py
from sqlalchemy import String
from sqlalchemy.orm import Mapped, mapped_column, DeclarativeBase

from .address import Address as Address

class User(DeclarativeBase):
    __tablename__ = "user"

    nome: Mapped[str] = mapped_column(String(50))
    addresses: Mapped[list["Address"]] = relationship(back_populates="users")

    ...

```
</details>

Having in mind that writing all the models in a single file is not an option...

Are there any workarounds that I can use?

**Comment by ntBre** [comment]
@adrianovieira I think that's a different issue from the one reported here and more closely related to https://github.com/astral-sh/ruff/issues/8700 and https://github.com/astral-sh/ruff/issues/7175. It looks like our current position is that this is a true positive, but #8700 tracks discussion around changing that decision.
